### PR TITLE
rel to #11520: introduce formula parser

### PR DESCRIPTION
**Pull request by eddiemuc**
rel to #11520: introduce formula parser

### CODE REVIEW

**Comment by baiti** [comment]
Don't know where to report this, whether it deserves a separate issue or perhaps there is also an existing issue:

When I create a waypoint and I go into "calculated coordinates" I use copy / paste today to get formulas into the "plain" field. If I then enter values for the variables in that formulas those won't be translated into coordinates if the entry field has leading and/or trailing white space, however, the correct coordinates are shown in the two lines above "Enter notes here".

The same is true if there is whitespace, for instance, before and/or after the dot that separates minutes from their fractions.

**Comment by eddiemuc** [comment]
@baiti I created issue #11586 for this

File: main/src/cgeo/geocaching/utils/formulas/FormulaTokenizer.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+import java.util.Set;
+
+import org.jetbrains.annotations.NotNull;
+
+public class FormulaTokenizer {
+
+    public enum Token { ID, NUMERIC, TEXT, SYMBOL, PAREN_OPEN, PAREN_CLOSE, PAREN_SEPARATOR, END }
+
+    private static final TokenData TOKENDATA_END = new TokenData(Token.END, null);
+    private static final char ESCAPE_CHAR = (char) 92; //backslash
```
[the remark this answers is not in the document]
Comment (eddiemuc) [reply main/src/cgeo/geocaching/utils/formulas/FormulaTokenizer.java:19]: i leave this as-is for now

**Comment by eddiemuc** [comment]
No longer relevant

**Pull request closed without merging**